### PR TITLE
feat: parallel pst opt

### PR DESCRIPTION
**Pull request by spetznick-elia**
## Checklist

Please check if the PR fulfills these requirements:

- [x] PR Title follows conventional commit messages
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] All commits in this PR are DCO signed-off (see CONTRIBUTING.md)

## Does this PR already have an issue describing the problem?

Fixes #

## What is the new behavior (if this is a feature change)?

This feature introduces grouping of controllable PSTs ("parallel PSTs") that enforces same tap changes for each group.
The optimization parameters in `StaticInformation` now include an option `enable_parallel_pst_group_optim` to use such groups during optimization.
PST ranges and groups are read from and persisted in the ActionSet. This means that if an ActionSet is available during importing, `ActionSet.pst_ranges[*].pst_group` will be used for grouping. If not ActionSet is found, the field will default to individual groups for each PST.

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

### CODE REVIEW

File: docs/dc_solver/preprocessing.md [hunk 1]
```diff
@@ -1,24 +1,46 @@
...
+The [`load_grid`][toop_engine_dc_solver.preprocess.convert_to_jax.load_grid] routine combines the latter two steps, runs an initial loadflow, and persists the standard solver artifacts back into the same processed grid folder.
 
 ## Data artifacts
 
-Output data pieces are defined in the [`folder_structure`][toop_engine_interfaces.folder_structure.PREPROCESSING_PATHS]. Most notably, the following data objects are written out:
+The processed grid folder layout is defined in the [`folder_structure`][toop_engine_interfaces.folder_structure.PREPROCESSING_PATHS]. The most important artifacts are split across the importer step and the DC solver step:
+
+| Stage | Artifact | Purpose |
+| --- | --- | --- |
+| Importer | `grid.xiidm` or `grid.json` | Backend-readable grid snapshot used by the powsybl or pandapower backend. |
```
Comment (BenjPetr) [inline docs/dc_solver/preprocessing.md:16]: I don't like the duplication. We should add the content of the interface here and have all the needed information in the interface written -> only one place to maintain the docs

File: packages/importer_pkg/src/toop_engine_importer/pypowsybl_import/powsybl_masks.py [hunk 2]
```diff
@@ -104,6 +105,14 @@ class NetworkMasks:
     trafo_pst_controllable: np.ndarray
     """Trafos which are a PST and can be controlled"""
 
+    pst_group_masks: np.ndarray
```
Comment (nicow-elia) [inline packages/importer_pkg/src/toop_engine_importer/pypowsybl_import/powsybl_masks.py]: At some point we should change the typing here to include shapes + dtypes as well. I would infer from the docstring Int[ndarray, " n_trafos"]? Or Int[ndarray, " n_trafos_with_phase_shift"]?

File: packages/interfaces_pkg/src/toop_engine_interfaces/backend.py [hunk 3]
```diff
@@ -296,6 +296,27 @@ def get_phase_shift_low_taps(self) -> Int[np.ndarray, " n_controllable_psts"]:
         """
         return np.zeros(sum(self.get_controllable_phase_shift_mask()), dtype=int)
 
+    def get_controllable_phase_shift_ids(self) -> list[str]:
```
Comment (nicow-elia) [inline packages/interfaces_pkg/src/toop_engine_interfaces/backend.py:299]: We should mention here that we never want this function to be overridden. Most of this interface is there to be overridden by later methods, so it is a bit confusing to suddenly have this method which, in my opinion, can not be implemented any different than this.

Alternatively you might make a helper function out of it, that would align more with the style we had so far.

File: packages/interfaces_pkg/src/toop_engine_interfaces/backend.py [hunk 4]
```diff
@@ -296,6 +296,27 @@ def get_phase_shift_low_taps(self) -> Int[np.ndarray, " n_controllable_psts"]:
...
+        branch_ids = self.get_branch_ids()
+        controllable_pst_mask = self.get_controllable_phase_shift_mask()
+        return [
+            str(branch_id)
+            for branch_id, is_controllable in zip(branch_ids, controllable_pst_mask, strict=True)
+            if is_controllable
+        ]
+
+    def get_parallel_pst_group_mask(self) -> Optional[Bool[np.ndarray, " n_parallel_pst_groups n_controllable_pst"]]:
+        """Get a PST group mask aligned with the controllable PST arrays.
```
Comment (nicow-elia) [inline packages/interfaces_pkg/src/toop_engine_interfaces/backend.py:311]: Link here in the docs to where a user who wants to implement this backend can find more information on the shape/design of this mask.
Alternatively copy+paste the docs here.

File: packages/interfaces_pkg/src/toop_engine_interfaces/backend.py [hunk 5]
```diff
@@ -296,6 +296,27 @@ def get_phase_shift_low_taps(self) -> Int[np.ndarray, " n_controllable_psts"]:
...
+
+        Returns None when no explicit grouping metadata is available.
+        """
+        return None
+
+    def get_parallel_pst_group_ids(self) -> Optional[list[str]]:
+        """Get PST group identifiers aligned with rows of get_parallel_pst_group_mask()."""
```
Comment (nicow-elia) [inline packages/interfaces_pkg/src/toop_engine_interfaces/backend.py]: Does the user have to override this when overriding the group masks? If yes, state so. If no, state what happens then.

File: packages/interfaces_pkg/src/toop_engine_interfaces/stored_action_set.py [hunk 6]
```diff
@@ -66,6 +66,19 @@ class PSTRange(GridElement):
     high_tap: int
     """The highest tap the PST supports"""
 
+    pst_group: str | None = None
+    """The optimization group of the PST.
+
+    When omitted in serialized action sets, this defaults to the PST id for backward compatibility.
```
Comment (nicow-elia) [inline packages/interfaces_pkg/src/toop_engine_interfaces/stored_action_set.py:72]: So it is never None?

File: packages/topology_optimizer_pkg/src/toop_engine_topology_optimizer/dc/genetic_functions/mutation/mutate_nodal_inj.py [hunk 7]
```diff
@@ -57,17 +69,32 @@ def mutate_psts(
...
-    # Keep the sample shape static so this function can run under vmap/jit.
-    mutation_samples = jax.random.normal(key_mutate, shape=pst_taps.shape) * pst_mutation_sigma
-    mutation = jnp.where(pst_indices_to_mutate, mutation_samples, 0.0)
-    mutation = jnp.round(mutation).astype(int)
-    new_pst_taps = pst_taps + mutation
-
-    # Reset random PSTs
-    pst_indices_to_reset = jax.random.bernoulli(key=key_reset, p=pst_reset_probability, shape=pst_taps.shape)
-    new_pst_taps = jnp.where(pst_indices_to_reset, pst_starting_taps, new_pst_taps)
+    if enable_parallel_pst_group_optim and parallel_pst_group_mask is not None:
```
Comment (nicow-elia) [inline packages/topology_optimizer_pkg/src/toop_engine_topology_optimizer/dc/genetic_functions/mutation/mutate_nodal_inj.py:72]: I would split this into two functions, seems they are doing very different things to me.

File: packages/dc_solver_pkg/src/toop_engine_dc_solver/preprocess/parallel_pst_groups.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        elif label in row_by_label:
+            rows[pst_index] = row_by_label[label]
+            continue
+        else:
+            row_by_label[label] = next_row
+            rows[pst_index] = next_row
+        group_ids.append(pst_id_list[pst_index])
+        next_row += 1
+
+    parallel_pst_group_mask = np.zeros((next_row, n_psts), dtype=bool)
```
Comment (nicow-elia) [inline packages/dc_solver_pkg/src/toop_engine_dc_solver/preprocess/parallel_pst_groups.py]: Don't use next_row as a shape here, just use len(group_lds) - more readable.

File: packages/dc_solver_pkg/src/toop_engine_dc_solver/preprocess/preprocess.py [hunk 9]
```diff
@@ -562,6 +562,16 @@ def reduce_branch_dimension(network_data: NetworkData) -> NetworkData:
     )
     relevant_phase_shift_starting_tap_idx = network_data.phase_shift_starting_tap_idx[kept_pst_branches]
     relevant_phase_shift_low_tap = network_data.phase_shift_low_tap[kept_pst_branches]
+    relevant_parallel_pst_group_mask = None
```
Comment (nicow-elia) [inline packages/dc_solver_pkg/src/toop_engine_dc_solver/preprocess/preprocess.py:565]: here and below in exclude_nonlinear_... you have added a logic to drop psts from the groups. Maybe these two logics could live in one function?

File: packages/dc_solver_pkg/src/toop_engine_dc_solver/jax/nodal_inj_optim.py [hunk 10]
```diff
@@ -37,6 +37,23 @@ def make_start_options(
     )
 
 
+def canonicalize_parallel_pst_taps(
+    pst_tap_indices: Int[Array, " batch_size n_timesteps n_controllable_pst"],
+    nodal_inj_info: NodalInjectionInformation,
+) -> Int[Array, " batch_size n_timesteps n_controllable_pst"]:
+    """Project PST taps onto the configured parallel-group constraint using one shared delta per group."""
```
Comment (nicow-elia) [inline packages/dc_solver_pkg/src/toop_engine_dc_solver/jax/nodal_inj_optim.py]: ... Which shared delta, the first, the last, the middle? Why are the taps "canonical" then and not just "grouped"? -> A bit more docs please :)

Comment (nicow-elia) [reply packages/dc_solver_pkg/src/toop_engine_dc_solver/jax/nodal_inj_optim.py]: Also describe why it is archicetcurally OK to find representative_indices here.

**Review by nicow-elia — COMMENTED** [review]
Mostly doc comments, looks good overall

File: packages/interfaces_pkg/src/toop_engine_interfaces/stored_action_set.py (hunk 6 above)
Comment (spetznick-elia) [reply packages/interfaces_pkg/src/toop_engine_interfaces/stored_action_set.py:72]: ~It is none when using the PandaPower backend (as we do not support this feature for PandaPower). I will clarify this here.~ Since I use after-validation the type annotation requires me to use str | None.

File: packages/importer_pkg/src/toop_engine_importer/pypowsybl_import/powsybl_masks.py (hunk 2 above)
Comment (spetznick-elia) [reply packages/importer_pkg/src/toop_engine_importer/pypowsybl_import/powsybl_masks.py]: agreed, the latter is correct

Comment (spetznick-elia) [reply packages/importer_pkg/src/toop_engine_importer/pypowsybl_import/powsybl_masks.py]: separate pr tho

File: packages/interfaces_pkg/src/toop_engine_interfaces/backend.py (hunk 3 above)
Comment (spetznick-elia) [reply packages/interfaces_pkg/src/toop_engine_interfaces/backend.py:299]: I changed the implementation such that this method is now abstract and the PandaPower backend merely returns None instead. Docs now reflect that we do not support parallel PST optimization for PandaPower.

Comment (spetznick-elia) [reply packages/interfaces_pkg/src/toop_engine_interfaces/backend.py:299]: I will also add a warning that if pandapower was used for importing that parallel PST optimization will not work.

File: packages/interfaces_pkg/src/toop_engine_interfaces/backend.py (hunk 5 above)
Comment (spetznick-elia) [reply packages/interfaces_pkg/src/toop_engine_interfaces/backend.py]: What to do when using pandapower moving forward is to be discussed

File: packages/topology_optimizer_pkg/src/toop_engine_topology_optimizer/dc/genetic_functions/mutation/mutate_nodal_inj.py (hunk 7 above)
Comment (spetznick-elia) [reply packages/topology_optimizer_pkg/src/toop_engine_topology_optimizer/dc/genetic_functions/mutation/mutate_nodal_inj.py:72]: resolved in new approach: groups always need to exist if `enable_parallel_pst_group_optim` is true

File: packages/dc_solver_pkg/src/toop_engine_dc_solver/jax/nodal_inj_optim.py (hunk 10 above)
Comment (spetznick-elia) [reply packages/dc_solver_pkg/src/toop_engine_dc_solver/jax/nodal_inj_optim.py]: removed this logic

**Comment by spetznick-elia** [comment]
the two last remaining bugs were found and addressed here: https://github.com/eliagroup/ToOp/pull/568

**Comment by spetznick-elia** [comment]
Closed in favor of #568
